gcsio/src/test/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystemNewIntegrationTest.java

```diff
@@ -677,6 +678,83 @@ public void delete_directory_parallel() throws Exception {
     delete_directory(/* parallelStatus= */ true);
   }
 
+  @Test
+  public void delete_inferred_directory() throws Exception {
+    GoogleCloudStorageOptions gcsOptions = GoogleCloudStorageOptions.builder()
+        .setAppName(GoogleCloudStorageTestHelper.APP_NAME)
+        .setAutoRepairImplicitDirectoriesEnabled(false)
+        .setProjectId(checkNotNull(TestConfiguration.getInstance().getProjectId())).build();
+
+    GoogleCloudStorageFileSystemOptions gcsFsOptions =
+        GoogleCloudStorageFileSystemOptions.builder().setCloudStorageOptions(gcsOptions)
+            .setStatusParallelEnabled(false).build();
+
+    TrackingHttpRequestInitializer gcsRequestsTracker =
+        new TrackingHttpRequestInitializer(httpRequestsInitializer);
+    GoogleCloudStorageFileSystem gcsFs = newGcsFs(gcsFsOptions, gcsRequestsTracker);
+
+    String bucketName = gcsfsIHelper.sharedBucketName1;
+    URI bucketUri = new URI("gs://" + bucketName + "/");
+    String dirObject = getTestResource();
+
+    gcsfsIHelper.createObjects(bucketName, dirObject + "/d1/d2");
+
+    gcsFs.delete(bucketUri.resolve(dirObject + "/d1/"), /* recursive= */ true);
+
+    assertThat(gcsRequestsTracker.getAllRequestStrings())
+        .containsExactly(
+            listRequestWithTrailingDelimiter(
+                bucketName, dirObject + "/d1/", /* maxResults= */ 1, /* pageToken= */ null),
+            listRequestString(
+                bucketName,
+                /* flatList= */ true,
+                /* includeTrailingDelimiter= */ null,
+                dirObject + "/d1/",
+                "bucket,name,generation",
+                /* maxResults= */ 1024,
+                /* pageToken= */ null),
+            deleteRequestString(bucketName, dirObject + "/d1/d2", /* generationId= */ 1));
+  }
+
+  @Test
+  public void delete_inferred_directory_with_parent() throws Exception {
+    GoogleCloudStorageFileSystemOptions gcsFsOptions =
+        newGcsFsOptions().setStatusParallelEnabled(false).build();
+
+    TrackingHttpRequestInitializer gcsRequestsTracker =
+        new TrackingHttpRequestInitializer(httpRequestsInitializer);
+    GoogleCloudStorageFileSystem gcsFs = newGcsFs(gcsFsOptions, gcsRequestsTracker);
+
+    String bucketName = gcsfsIHelper.sharedBucketName1;
+    URI bucketUri = new URI("gs://" + bucketName + "/");
+    String dirObject = getTestResource();
+
+    gcsfsIHelper.createObjects(bucketName, dirObject + "/d1/d2/d3");
+
+    gcsFs.delete(bucketUri.resolve(dirObject + "/d1/d2"), /* recursive= */ true);
+
+    assertThat(gcsFs.exists(bucketUri.resolve(dirObject + "/d1/"))).isTrue();
```

Comment:
Can you use `gcsfsIHelper` to check objects existence? Or move this check after GCS requests assertion because right now it asserts requests for this check too.

Reply:
Done